Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm

```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
+{
+    Vector<WebKit::WebsiteDataType> dataTypesToEncode;
+    for (NSString *dataType in dataTypes) {
+        if ([dataType isEqualToString:WKWebsiteDataTypeLocalStorage]) {
+            dataTypesToEncode.append(WebKit::WebsiteDataType::LocalStorage);
+            continue;
+        }
+
+        completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:NSFeatureUnsupportedError userInfo:NULL]);
+    }
+
+    auto data = Box<WKWebsiteData>::create();
+
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), dataTypesToEncode = WTFMove(dataTypesToEncode), data] {
+        WTF::Persistence::Encoder encoder;
+        constexpr unsigned currentWKWebsiteDataSerializationVersion = 1;
+        encoder << currentWKWebsiteDataSerializationVersion;
+        encoder << dataTypesToEncode;
+
+        for (auto& dataTypeToEncode : dataTypesToEncode) {
+            switch (dataTypeToEncode) {
+            case WebKit::WebsiteDataType::LocalStorage:
+                if (!data->localStorage) {
+                    completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
+                    return;
+                }
+                encoder << data->localStorage.value();
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
+            }
+        }
+
+        completionHandler(toNSData(encoder.span()).get(), NULL);
+    });
+
+    if ([dataTypes containsObject:WKWebsiteDataTypeLocalStorage]) {
+        _websiteDataStore->fetchLocalStorage([callbackAggregator, data](auto&& localStorage) {
+            data->localStorage = WTFMove(localStorage);
+        });
+    }
+}
+
+- (void)restoreData:(NSData *)data completionHandler:(void(^)(NSError *))completionHandler
+{
+    WTF::Persistence::Decoder decoder(span(data));
+
+    std::optional<unsigned> currentWKWebsiteDataSerializationVersion;
+    decoder >> currentWKWebsiteDataSerializationVersion;
+    if (!currentWKWebsiteDataSerializationVersion) {
+        completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:NULL]);
+        return;
+    }
+
+    std::optional<Vector<WebKit::WebsiteDataType>> encodedDataTypes;
+    decoder >> encodedDataTypes;
+    if (!encodedDataTypes) {
+        completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:NULL]);
+        return;
+    }
+
+    auto succeeded = Box<bool>::create(true);
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), succeeded] {
+        if (*succeeded)
+            completionHandler(NULL);
+        else
+            completionHandler([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```

Comment:
I'm sure this is not the right error but am not sure what would be better. This will be called if the Network Process crashes and the completion handler is called with the default arguments (std::nullopt). Any suggestions?